diff --git a/ruby/lib/ruby/gems/1.9.1/gems/win32-dir-0.4.3/lib/win32/dir.rb b/ruby/lib/ruby/gems/1.9.1/gems/win32-dir-0.4.3/lib/win32/dir.rb
index 815019e..563927e 100644
--- a/ruby/lib/ruby/gems/1.9.1/gems/win32-dir-0.4.3/lib/win32/dir.rb
+++ b/ruby/lib/ruby/gems/1.9.1/gems/win32-dir-0.4.3/lib/win32/dir.rb
@@ -88,6 +88,7 @@ class Dir
       if SHGetFolderLocation(0, value, 0, 0, ptr) == 0
         if SHGetFileInfo(ptr.read_long, 0, info, info.size, flags) != 0
           path = info[:szDisplayName].to_s
+          path.force_encoding(Encoding.default_external)
         end
       end
     end
